common/lib/client/resource.ts

```diff
@@ -0,0 +1,173 @@
+		}
+		if (callback) { callback(err, body, headers, unpacked, statusCode); }
+	}
+}
+
+class Resource {
+	static get(rest: Rest, path: string, origheaders: Record<string, string>, origparams: Record<string, any>, envelope: Utils.Format | null, callback: Function): void {
+		Resource.do(HttpMethods.Get, rest, path, null, origheaders, origparams, envelope, callback);
+	} 
+
+	static delete(rest: Rest, path: string, origheaders: Record<string, string>, origparams: Record<string, any>, envelope: Utils.Format | null, callback: Function): void {
+		Resource.do(HttpMethods.Delete, rest, path, null, origheaders, origparams, envelope, callback);
+	} 
+
+	static post(rest: Rest, path: string, body: unknown, origheaders: Record<string, string>, origparams: Record<string, any>, envelope: Utils.Format | null, callback: Function): void {
+		Resource.do(HttpMethods.Post, rest, path, body, origheaders, origparams, envelope, callback);
+	} 
+
+	static patch(rest: Rest, path: string, body: unknown, origheaders: Record<string, string>, origparams: Record<string, any>, envelope: Utils.Format | null, callback: Function): void {
+		Resource.do(HttpMethods.Patch, rest, path, body, origheaders, origparams, envelope, callback);
+	} 
+
+	static put(rest: Rest, path: string, body: unknown, origheaders: Record<string, string>, origparams: Record<string, any>, envelope: Utils.Format | null, callback: Function): void {
+		Resource.do(HttpMethods.Put, rest, path, body, origheaders, origparams, envelope, callback);
+	} 
+
+	static do(method: HttpMethods, rest: Rest, path: string, body: unknown, origheaders: Record<string, string>, origparams: Record<string, any>, envelope: Utils.Format | null, callback: Function): void {
+		if (Logger.shouldLog(Logger.LOG_MICRO)) {
+			callback = logResponseHandler(callback, method, path, origparams);
+		}
+
+		if(envelope) {
+			callback = (callback && unenvelope(callback, envelope));
+			(origparams = (origparams || {}))['envelope'] = envelope;
+		}
+
+		function doRequest(this: any, headers: Record<string, string>, params: Record<string, any>) {
+			if (Logger.shouldLog(Logger.LOG_MICRO)) {
+				Logger.logAction(Logger.LOG_MICRO, 'Resource.' + method + '()', 'Sending; ' + urlFromPathAndParams(path, params));
+			}
+
+			const args = [rest, path, headers, body, params, function(err: ErrorInfo, res: any, headers: Record<string, string>, unpacked: boolean, statusCode: number) {
+				if(err && Auth.isTokenErr(err)) {
+					/* token has expired, so get a new one */
+					rest.auth.authorize(null, null, function(err: Error) {
+						if(err) {
+							callback(err);
+							return;
+						}
+						/* retry ... */
+						withAuthDetails(rest, origheaders, origparams, callback, doRequest);
+					});
+					return;
+				}
+				callback(err, res, headers, unpacked, statusCode);
+			}];
+			if (!body) {
+				args.splice(3, 1);
```

Comment:
I think this one deserves an explanation comment :)

Reply:
7d8de2ba

Reply:
Resolving this conversation on @ikbalkaya's behalf.